### PR TITLE
Fixing the omitempty tag

### LINKED ISSUE #7730: API: Fix omitempty tags
Some of the API fields have incorrect omitempty tags, which needs to be fixed.
For ex: imagePullPolicy in Container should be omitempty, but it is not.

@madhuri-rai07

### PR DESCRIPTION
**Pull request by nikhiljindal**
Fixes https://github.com/GoogleCloudPlatform/kubernetes/issues/7730 and https://github.com/GoogleCloudPlatform/kubernetes/issues/6841

Fixing the omitempty tag for v1beta3 and v1 APIs.

### CODE REVIEW

**Comment by nikhiljindal** [comment]
@bgrant0607

File: pkg/api/v1/types.go [hunk 1]
```diff
@@ -435,7 +435,7 @@ type AWSElasticBlockStoreVolumeSource struct {
 	// Ex. "ext4", "xfs", "ntfs"
 	// TODO: how do we prevent errors in the filesystem from compromising the machine
 	// TODO: why omitempty if required?
```
Comment (bgrant0607) [inline pkg/api/v1/types.go]: Remove TODO

Comment (nikhiljindal) [inline pkg/api/v1/types.go]: Removed.

File: pkg/api/v1beta3/types.go [hunk 2]
```diff
@@ -200,26 +200,26 @@ type VolumeSource struct {
 	// to see the host machine. Most containers will NOT need this.
```
Comment (bgrant0607) [inline pkg/api/v1beta3/types.go]: On https://github.com/nikhiljindal/kubernetes/blob/omitEmpty/pkg/api/v1beta3/types.go#L191
```VolumeSource `json:",inline,omitempty"````
inline fields shouldn't have omitempty

Comment (bgrant0607) [inline pkg/api/v1beta3/types.go]: Also, everywhere, except PodTemplateSpec and Binding, ObjectMeta should not be tagged with omitempty.

File: pkg/api/v1beta3/types.go [hunk 3]
```diff
@@ -643,7 +642,7 @@ type Container struct {
 	// Deprecated - see SecurityContext.  Optional: Default to false.
 	Privileged bool `json:"privileged,omitempty" description:"whether or not the container is granted privileged status; defaults to false; cannot be updated; deprecated;  See SecurityContext."`
 	// Optional: Policy for pulling images for this container
-	ImagePullPolicy PullPolicy `json:"imagePullPolicy" description:"image pull policy; one of PullAlways, PullNever, PullIfNotPresent; defaults to PullAlways if :latest tag is specified, or PullIfNotPresent otherwise; cannot be updated"`
+	ImagePullPolicy PullPolicy `json:"imagePullPolicy,omitempty" description:"image pull policy; one of PullAlways, PullNever, PullIfNotPresent; defaults to PullAlways if :latest tag is specified, or PullIfNotPresent otherwise; cannot be updated"`
 	// Deprecated - see SecurityContext.  Optional: Capabilities for container.
 	Capabilities Capabilities `json:"capabilities,omitempty" description:"capabilities for container; cannot be updated; deprecated; See SecurityContext."`
```
Comment (bgrant0607) [inline pkg/api/v1beta3/types.go]: PodSpec.Volumes should have omitempty

File: pkg/api/v1beta3/types.go [hunk 4]
```diff
@@ -1010,7 +1009,7 @@ type ServicePort struct {
 	// If this is a string, it will be looked up as a named port in the
 	// target Pod's container ports.  If this is not specified, the value
 	// of Port is used (an identity map).
-	TargetPort util.IntOrString `json:"targetPort" description:"the port to access on the pods targeted by the service; defaults to the service port"`
+	TargetPort util.IntOrString `json:"targetPort,omitempty" description:"the port to access on the pods targeted by the service; defaults to the service port"`
```
Comment (bgrant0607) [inline pkg/api/v1beta3/types.go]: Another github annoyance -- I can't comment on arbitrary lines. :-(

ServiceAccount.Secrets should have omitempty:
https://github.com/nikhiljindal/kubernetes/blob/omitEmpty/pkg/api/v1beta3/types.go#L1052

File: pkg/api/v1beta3/types.go [hunk 5]
```diff
@@ -1321,7 +1320,7 @@ type DeleteOptions struct {
 	// Optional duration in seconds before the object should be deleted. Value must be non-negative integer.
 	// The value zero indicates delete immediately. If this value is nil, the default grace period for the
 	// specified type will be used.
-	GracePeriodSeconds *int64 `json:"gracePeriodSeconds" description:"the duration in seconds to wait before deleting this object; defaults to a per object value if not specified; zero means delete immediately"`
+	GracePeriodSeconds *int64 `json:"gracePeriodSeconds,omitempty" description:"the duration in seconds to wait before deleting this object; defaults to a per object value if not specified; zero means delete immediately"`
 }
```
Comment (bgrant0607) [inline pkg/api/v1beta3/types.go]: Again, not this line.

All the ListOptions should have have omitempty.
https://github.com/nikhiljindal/kubernetes/blob/omitEmpty/pkg/api/v1beta3/types.go#L1331

**Comment by bgrant0607** [comment]
Thanks much. I reviewed the whole v1beta3 file. Please make equivalent changes to v1.

File: pkg/api/v1beta3/types.go (hunk 2 above)
Comment (nikhiljindal) [inline pkg/api/v1beta3/types.go]: Why is ObjectMeta not omitempty? All the fields in ObjectMeta are omitempty.
Not making it omitempty will just force users to include "metadata:{}" in their configs.

Comment (nikhiljindal) [inline pkg/api/v1beta3/types.go]: Updated VolumeSource

File: pkg/api/v1beta3/types.go (hunk 3 above)
Comment (nikhiljindal) [inline pkg/api/v1beta3/types.go]: Done

File: pkg/api/v1beta3/types.go (hunk 4 above)
Comment (nikhiljindal) [inline pkg/api/v1beta3/types.go]: Done

File: pkg/api/v1beta3/types.go (hunk 5 above)
Comment (nikhiljindal) [inline pkg/api/v1beta3/types.go]: Done

**Comment by nikhiljindal** [comment]
Updated both v1beta3 and v1 as per comments, modulo one question regarding ObjectMeta.

File: pkg/api/v1beta3/types.go (hunk 2 above)
Comment (bgrant0607) [inline pkg/api/v1beta3/types.go]: Users have to specify either name or generateName, in post, put, and patch, and it will be returned by get.

Comment (bgrant0607) [inline pkg/api/v1beta3/types.go]: It's ok to leave them all omitempty for now, but I can't think of a scenario where metadata would be absent. We could maybe extract the name from the URLs on some operations, but I don't think we do that currently.

**Comment by bgrant0607** [comment]
LGTM

**Pull request merged**